src/system/SystemLayerImplOpenIoTSDK.cpp

```diff
@@ -0,0 +1,441 @@
+            }
+            iotSocketMaskSet(socket_id, GetSelectMask(ReadMask));
+        }
+    }
+
+    // if we failed we need to close the socket to be able to retry
+    if (err != CHIP_NO_ERROR)
+    {
+        iotSocketClose(socket_id);
+    }
+    else
+    {
+        mSignalSocket = socket_id;
+    }
+
+    mSelectMutex.Unlock();
+
+    return err;
+}
+
+CHIP_ERROR LayerImplOpenIoTSDK::EnableSelectCallback(chip::Inet::EndPointStateIoTSocket * endpoint, bool read, bool write)
+{
+    EnableSignalSocket();
+    mSelectMutex.Lock();
+
+    bool add = true;
+    for (size_t i = 0; i < mSelectEndpointsNumber; ++i)
+    {
+        if (mSelectEndpoints[i] == endpoint)
+        {
+            add = false;
+            // already enabled
+            break;
+        }
+    }
+
+    if (add)
+    {
+        if (mSelectEndpointsNumber == kSocketMax)
+        {
+            // impossible to have more endpoints than allowed
+            // maybe some endpoints haven't removed their callback
+            mSelectMutex.Unlock();
+            return CHIP_ERROR_INCORRECT_STATE;
+        }
+        else
+        {
+            mSelectEndpoints[mSelectEndpointsNumber] = endpoint;
+            mSelectEndpointsNumber++;
+        }
+    }
+
+    if (read)
+    {
+        iotSocketMaskSet(endpoint->GetSocketId(), GetSelectMask(ReadMask));
+    }
+    if (write)
+    {
+        iotSocketMaskSet(endpoint->GetSocketId(), GetSelectMask(WriteMask));
+    }
+
+    mSelectMutex.Unlock();
+
+    Signal();
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR LayerImplOpenIoTSDK::DisableSelectCallback(chip::Inet::EndPointStateIoTSocket * endpoint, bool read, bool write)
+{
+    EnableSignalSocket();
+    mSelectMutex.Lock();
+
+    // if the either mask remains set do not remove
+    bool keep = false;
+    if (read)
+    {
+        iotSocketMaskUnset(endpoint->GetSocketId(), GetSelectMask(ReadMask));
+    }
+    else
+    {
+        keep = iotSocketMaskIsSet(endpoint->GetSocketId(), GetSelectMask(ReadMask));
+    }
+
+    if (write)
+    {
+        iotSocketMaskUnset(endpoint->GetSocketId(), GetSelectMask(WriteMask));
+    }
+    else
+    {
+        keep |= iotSocketMaskIsSet(endpoint->GetSocketId(), GetSelectMask(WriteMask));
+    }
+
+    if (!keep)
+    {
+        // find the one to remove
+        for (size_t i = 0; i < mSelectEndpointsNumber; ++i)
+        {
+            if (mSelectEndpoints[i] == endpoint)
+            {
+                mSelectEndpoints[i] = nullptr;
+                mSelectEndpointsNumber--;
+
+                // if this element wasn't last, grab the last one which is now beyond mSelectEndpointsNumber limit
+                if (i != mSelectEndpointsNumber)
+                {
+                    mSelectEndpoints[i] = mSelectEndpoints[mSelectEndpointsNumber];
+                }
+                break;
+            }
+        }
+    }
+
+    mSelectMutex.Unlock();
+
+    mSignalMutex.Lock();
+    osSemaphoreAcquire(mSignalSem, osWaitForever);
```

Comment:
this is a workaround the LWIP bug where select does not wake on socket close resulting in assert if we open another socket that lwip assigns the reused ID which we are already waiting on

Reply:
@pan-